docs: add 'What Flint Does Differently' blog post (#1208)
flint-fyi/flint: packages/site/src/content/docs/blog/what-flint-does-differently.mdx

- Core rules are less powerful than they could be
- Custom rules have to choose between being fast and easy to set up vs. slower and type-aware
- Even if the core architecture supported typed rules, it'd be extra work to make separate code paths
- Users now have to keep track of >=1 extra concept while configuring their linter

Given Flint's hybrid core, typed linting should be significantly faster and much easier to set up than in ESLint.
Removing typed linting's downsides simplifies the linting story:

- Core rules don't need to be duplicated by plugins to add in typed linting support
- Custom rules are simpler to write for not having to make decisions around types
- The core linter architecture can be optimized for type-checked linting performance
- Users no longer need to care about concepts like typed linting while configuring their linter

In building type awareness into the core runtime, Flint will by necessity also need to support stateful languages like TypeScript.
This will solve the common ["ESLint does not re-compute cross-file information on file changes" issue for editors](https://github.com/microsoft/vscode-eslint/issues/1774).

### Built-In TypeScript Support

> Flint provides TypeScript support out-of-the-box, the way Biome, deno lint Oxlint, and other recent linters all do.

Reviewer: ```suggestion
Flint provides TypeScript support out-of-the-box, the way Biome, deno_lint, Oxlint, and other recent linters all do.
```

Reviewer: TL;DR: It's both, but you win.

Ok, had to go digging because yes, those docs call it `deno lint` but I've _mostly_ seen it referred to as deno_lint (tbh I can't find sources for that anymore, at bare minimum I think Biome used to call it deno_lint in the comparisons page, but they call it Deno Lint now). I was about to give up and announce I was crazy... and then I found it: [denoland/deno_lint](https://github.com/denoland/deno_lint). `deno_lint` is the crate, `deno lint` is the CLI subcommand that embeds `deno_lint`. Given it's that ambiguous and everybody else seems to gravitated toward deno lint, I'm happy to let it be. But Deno should be capitalized at least.

```suggestion
Flint provides TypeScript support out-of-the-box, the way Biome, Deno lint, Oxlint, and other recent linters all do.
```

### Core Common Languages

Building in TypeScript support is a great start, but linting is useful for more than just JavaScript and TypeScript.
ESLint plugins exist for basically every language that web repositories use.
We believe a modern web linter should encourage applying the same developer assistance and quality checks to all the languages in a project.

Flint additionally provides linting for JSON, Markdown, and YAML out-of-the-box.
No additional plugins needed.
Some other linters are already moving in this direction:

- [Biome supports a plethora of frontend languages](https://biomejs.dev/internals/language-support) with [YAML support planned](https://github.com/biomejs/biome/issues/2365 "Biome issue#2365 YAML support")
- ESLint: already provides [`@eslint/json`](https://github.com/eslint/json) and [`@eslint/markdown`](https://github.com/eslint/markdown)

Flint's core architecture is completely agnostic of any specific language.
Each language provides its own parser, type information services, and any other language-specific hooks.
Users are of course still be able to write their own plugins for other languages.

### Focused Language Plugins

> Flint will additionally provide first-party _plugins_ for common languages: CSS, JSX, Vue, and so on.

Reviewer: Yeah, thus my comment that:

> I don't know that HTML support is worthwhile

Listing CSS makes me think JS/CSS/HTML, but linting-scale projects don't really use HTML, so it's only a useful language if it's able to lint over embedded languages.

Flint aims to provide a consistent and high quality user experience for these languages by building them in the Flint project.

### Formatting Coordination

[Formatters are not linters](https://typescript-eslint.io/users/what-about-formatting/#formatters-vs-linters).
Two separate tools should be used for those two separate concerns.
However, _configuring_ two tools is cumbersome for users - especially given that the list of files you'd want to _format_ is almost always roughly the same as the list of files to _lint_.

One of the biggest reasons users move to Biome is that those it performs both _formatting_ and _linting_ with a single devDependency and configuration file.
In doing so, they provide a much easier setup and maintenance story, as well as sidestep many common ESLint misconfigurations that lead to performance issues.

Flint embraces the Biome approach of having one tool _coordinate_ others.
Flint will automatically run [Prettier](https://prettier.io) on linted files to format them after linting.

:::tip
See more details on this point in [Introducing Flint > Tooling Coordination](/blog/introducing-flint#tooling-coordination).
:::

### Language Diagnostics Coordination

> As with formatting, most users generally run _type checking_ on roughly the same set of files as linting.

Author: > If that's desired, we should note it as a trade-off.

I see it as a desirable trade-off. IMO if you want to type-check the file, you should also lint it. (there is nuance there we should probably discuss... somewhere)

> does this mean Flint coordinates `vue-tsc`, `mdx-analyzer`, `astro check`, etc?

I'm not familiar personally with the first two, but if they're language-level reports then perhaps?

`astro check` I believe so, though the byproduct of running `astro sync` wouldn't be doable.

> As with formatting, most users generally run _type checking_ on roughly the same set of files as linting.

Reviewer 2: > Also, I don't know how well auvred's experiments are working, but does this mean Flint coordinates `vue-tsc`, `mdx-analzyer` (which doesn't even have a CLI yet), `astro check`, etc? If it does, that's a real win for multilingual projects which we should note _somewhere_. Probably not here.

`vue-tsc` and `astro check` both do the following:

- Write a Volar.js language plugin (the most difficult part - basically translating source `.vue`/`.astro` files to `.ts`/`.tsx` with Volar mappings, etc.)
- Initialize the Volar.js language with this language plugin
- [Patch the TS program](https://github.com/volarjs/volar.js/blob/574c2ceea001bf359557508def5913da3c27b7f1/packages/typescript/lib/quickstart/runTsc.ts) and run `tsc` (in the case of `vue-tsc`) or [run custom Volar.js checker](https://github.com/volarjs/volar.js/blob/574c2ceea001bf359557508def5913da3c27b7f1/packages/kit/lib/createChecker.ts) (in the case of `astro check`)

Currently, we follow the same route as `vue-tsc` because we're using the TS Project Service. It has a [very](https://github.com/microsoft/TypeScript/blob/f3770c9b89ba7969b163c5cea7805d5df23b2086/src/server/editorServices.ts), [very](https://github.com/microsoft/TypeScript/blob/f3770c9b89ba7969b163c5cea7805d5df23b2086/src/server/project.ts) large amount of ~~code~~ semantic usefulness for us, so we definitely don't want to rewrite it ourselves. Although [Volar.js uses a custom project service in its language server](https://github.com/volarjs/volar.js/tree/574c2ceea001bf359557508def5913da3c27b7f1/packages/language-server/lib/project).

I couldn't find a way to customize language service for TS Projects created by TS Project Service without patching the `typescript` (I do want to explore runtime prototype patching of the `ts.Project` class, though. My guess is that if it works the way I expect, we could get rid of the `ts-patch` package.)

So what we do is:

- Patch `typescript` package code (register `.astro` and `.vue` extensions in global hardcoded lists, and set TS program creation proxy)
- Let `@flint.fyi/astro-language` and `@flint.fyi/vue-language` register their Volar language plugins globally
- Create TS Project Service
- `ts.ProjectService` creates a `ts.Project`, and it creates a`ts.Program`
- We intercept `ts.Program` creation and create Volar language using the Astro and Vue language plugins
- Now our `ts.Program` has `.ts`, `.astro`, `.vue` files, and we can collect syntactic & semantic diagnotics from all of them at once!

So, yeah, probably one more killer feature!

We can even typecheck `.astro` and `.vue` files that are not included in `tsconfig.json`, since the Project Service creates a default project for such files.

> As with formatting, most users generally run _type checking_ on roughly the same set of files as linting.

Reviewer: > Just a general comment, AFAIK we don't typecheck unlinted files, so if users had, e.g. a generated file ignored from linting, that'd ignore type errors?

Just found this old comment: https://github.com/flint-fyi/flint/pull/141#issuecomment-3058212903. So to answer my original question: yes and yes.

> As with formatting, most users generally run _type checking_ on roughly the same set of files as linting.

Author: This is awesome!! Yes I think that'd be an awesome feature. As a long-time Astro + MDX user I would definitely appreciate integrated linting & type checking for them, it's kind of a pain. Even with Astro setting everything up so nicely for users, having multiple commands / places to look is a pain.

      - `enforce*`: [`accessor-pairs` > `enforceForClassMembers`](https://eslint.org/docs/latest/rules/accessor-pairs#enforceforclassmembers)
      - `include*`: [`no-duplicate-imports` > `includeExports`](https://eslint.org/docs/latest/rules/no-duplicate-imports#options)
      - `require*`: [`arrow-parens` > `requireForBlockBody`](https://eslint.style/rules/js/arrow-parens#requireforblockbody)
- Value defaults:
  - Boolean values:
    - Off-by-default: [`no-floating-promises` > `ignoreIIFE`](https://typescript-eslint.io/rules/no-floating-promises#ignoreiife)
    - On-by-default: [`no-floating-promises` > `ignoreVoid`](https://typescript-eslint.io/rules/no-floating-promises#ignorevoid)
  - Multi-values:
    - Empty by default: [`no-floating-promises` > `allowForKnownSafeCalls`](https://typescript-eslint.io/rules/no-floating-promises#allowforknownsafecalls)
    - Starting set: [`restrict-template-expressions` > `allow`](https://typescript-eslint.io/rules/restrict-template-expressions#allow)

What a list!

> Flint's rules attempt make the same choice consistently for their options:

Reviewer: Yeah, that sounds better!

| [`ignoreIIFE`](https://typescript-eslint.io/rules/no-floating-promises#ignoreiife)                                                       | _(same)_             | `false`         | _(same)_         |
| [`ignoreVoid`](https://typescript-eslint.io/rules/no-floating-promises#ignorevoid)                                                       | _(same)_             | `true`          | `false`          |
| [`includeExports`](https://eslint.org/docs/latest/rules/no-duplicate-imports#options)                                                    | `checkExports`       | `false`         | _(same)_         |
| [`requireForBlockBody`](https://eslint.style/rules/js/arrow-parens#requireforblockbody)                                                  | `checkBlockBodies`   | `false`         | _(same)_         |
| [`skipBlankLines`](https://eslint.style/rules/js/no-trailing-spaces#skipblanklines)                                                      | `ignoreBlankLines`   | `false`         | _(same)_         |

Flint will also sort rule options alphabetically in their docs pages.
Users should be able to O(log(N)) scan a docs page for a rule option.

> Not O(1) search through some arbitrary order.

Reviewer: I'm not an expert on big-O notation, but wouldn't it be O(n)? (if not I'd like to learn)

```suggestion
Not O(n) search through some arbitrary order.
```

Author: Oops, yes. You got it right. I didn't.

You don't get editor intellisense while authoring; instead, you have to use [`@eslint/config-inspector`](https://www.npmjs.com/package/@eslint/config-inspector) or run your config to know whether you've mistyped the name of a rule or an option.

Flint's rules will take a more modern approach.
They will allow [Standard Schema](https://standardschema.dev) descriptions for rule options, which allows TypeScript-friendly schema validations libraries such as [Zod](https://zod.dev):

```ts
import { typescriptLanguage } from "@flint/ts";
import { z } from "zod";

export const someRule = typescriptLanguage.createRule({
	options: {
		option: z.string(),
	},
	setup(context) {
		// ...
	},
});
```

```ts title="flint.config.ts"
import { defineConfig, ts } from "flint";
import { someRule } from "./src/rules/someRule.ts";

export default defineConfig({
	use: [
		{
			files: ts.files.all,
			rules: [
				someRule({
					option: "Hooray!",
				}),
			],
		},
	],
});
```

As a result, Flint's configuration is _fully typed_.

> Users receive intellisense as they type plugin rules, and all those settings can be type checked.

Author: I would _love_ to see that lawsuit 😂

:::tip
See more details on this point in [Introducing Flint > Thorough Provided Plugins](/blog/introducing-flint#thorough-provided-plugins).
:::

### Community Organization

The [ESLint Community organization](https://github.com/eslint-community) is wonderful.
It serves a great need for housing high-applicability, high-value community projects that are not able to be part of ESLint core.
It's a kind of "next step" for finding plugins outside of ESLint core -- not quite "first party", and not an external "third party".

Flint will lean into having an equivalent community organization.
That organization would have guidelines for inclusion, including:

- Actively supporting new core linter minor versions soon after release
- Adherence to the shared linting glossary
- Documenting all configs, rules, and rule options
- Maximum time to (re-)triage issues and pull requests
- Minimum numbers of consuming projects and weekly downloads
- Not being specific to any one userland framework
- Providing metadata alongside the `package.json` such as:
  - Names of any dependencies the plugin is directly for (e.g. `"lodash"`, `"react"`)

>   - Text of when to use the plugin and each of its configs

Reviewer: I mean, this would clear it up for me given we _do_ intent to keep Vue, Astro, Svelte, languages in core, just not code for their frameworks, but perhaps you meant it more broadly? (as in niche, except that I (unfortunately) can't call Next.js niche?)

```suggestion
- Not being specific to any one userland metaframework
- Providing metadata alongside the `package.json` such as:
  - Names of any dependencies the plugin is directly for (e.g. `"lodash"`, `"react"`)
  - Text of when to use the plugin and each of its configs
```

Flint's "incubator" plugins are each intended to eventually grow into those community projects.

### Plugin Registry

> Much of a linter's ecosystem will always be third-party plugins.

Reviewer: How about:

```suggestion
A linter's ecosystem will always include a fair share of third-party plugins.
```

> Well, with Flint, that might not be true anymore!

I interpreted the current wording as saying that 3rd party plugins will always be the most important part of an ecosystem 😅, so this definitely needs changing then.
